plugins/destination/duckdb/client/write.go

```diff
@@ -134,39 +134,46 @@ func (c *Client) WriteTableBatch(ctx context.Context, table *schema.Table, recor
-			return err
-		}
-	} else {
-		tmpTableName := table.Name + strings.ReplaceAll(uuid.New().String(), "-", "_")
-		if err := c.createTableIfNotExist(ctx, tmpTableName, table); err != nil {
-			return fmt.Errorf("failed to create table %s: %w", tmpTableName, err)
-		}
-		if err := c.copyFromFile(ctx, tmpTableName, f.Name(), sc); err != nil {
-			return fmt.Errorf("failed to copy from file %s: %w", f.Name(), err)
-		}
+	if !c.enabledPks() || len(table.PrimaryKeys()) == 0 {
+		return c.copyFromFile(ctx, table.Name, f.Name(), sc)
+	}
 
-		// At time of writing (March 2023), duckdb does not support updating list columns.
-		// As a workaround, we delete the row and insert it again. This makes it non-atomic, unfortunately,
-		// but this is unavoidable until support is added to duckdb itself.
-		// See https://github.com/duckdb/duckdb/blob/c5d9afb97bbf0be12216f3b89ae3131afbbc3156/src/storage/table/list_column_data.cpp#L243-L251
-		if containsList(table) {
-			if err := c.deleteByPK(ctx, tmpTableName, table.Name, table); err != nil {
-				return err
-			}
-			if _, err = c.db.ExecContext(ctx, "insert into "+table.Name+" from "+tmpTableName); err != nil {
-				return fmt.Errorf("failed to insert into %s from %s: %w", table.Name, tmpTableName, err)
-			}
-		} else {
-			if err := c.upsert(ctx, tmpTableName, table.Name, table); err != nil {
-				return err
-			}
-		}
-		if _, err = c.db.ExecContext(ctx, "drop table "+tmpTableName); err != nil {
-			return err
+	tmpTableName := table.Name + strings.ReplaceAll(uuid.New().String(), "-", "_")
+	if err := c.createTableIfNotExist(ctx, tmpTableName, table); err != nil {
+		return fmt.Errorf("failed to create table %s: %w", tmpTableName, err)
+	}
+	defer func() {
+		e := c.exec(ctx, "drop table "+tmpTableName)
+		if err == nil {
+			// we preserve original error, so update only on nil err
+			err = e
 		}
+	}()
+	if err := c.copyFromFile(ctx, tmpTableName, f.Name(), sc); err != nil {
+		return fmt.Errorf("failed to copy from file %s: %w", f.Name(), err)
 	}
 
-	return nil
+	// At time of writing (March 2023), duckdb does not support updating list columns.
+	// As a workaround, we delete the row and insert it again. This makes it non-atomic, unfortunately,
+	// but this is unavoidable until support is added to duckdb itself.
+	// See https://github.com/duckdb/duckdb/blob/c5d9afb97bbf0be12216f3b89ae3131afbbc3156/src/storage/table/list_column_data.cpp#L243-L251
+	if containsList(table) {
+		return c.deleteInsert(ctx, tmpTableName, table)
+	}
+
+	return c.upsert(ctx, tmpTableName, table)
+}
+
+func (c *Client) deleteInsert(ctx context.Context, tmpTableName string, table *schema.Table) error {
+	if err := c.deleteByPK(ctx, tmpTableName, table); err != nil {
+		return err
+	}
+
+	// per https://duckdb.org/docs/sql/indexes#over-eager-unique-constraint-checking we might need to retry
+	return backoff.Retry(
+		func() error {
+			return c.exec(ctx, "insert into "+table.Name+" from "+tmpTableName)
```

Comment:
is there any way we can detect the unique constraint violation error and only retry that, rather than retrying all errors?

Reply:
We could detect it only via comparing error strings, but I'd rather leave it to the further implementation if we ever need to touch this place once again.
I hope that the [issue](https://duckdb.org/docs/sql/indexes#over-eager-unique-constraint-checking) is fixed by DuckDB itself thus making this code obsolete and enabling us to remove this backoff sooner though.